src/form-validation-summary-builder.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  ValidationResult,
+  FormValidationSummary,
+  createDefaultFormValidationSummary,
+  FormFieldError,
+} from './model';
+import { arrayContainsElements } from './helper';
+import { globalFormValidationId } from './const';
+
+const didAllValidationsSucceeded = (
+  validationResults: ValidationResult[]
+): boolean => validationResults.every(fvr => fvr.succeeded);
+
+const extractFieldErrors = (
+  validationResults: ValidationResult[]
+): FormFieldError[] => {
+  const fieldValidationResults = validationResults.filter(
+    fvr => fvr.key !== globalFormValidationId
+  );
+
+  return fieldValidationResults.map(validation => ({
+    key: validation.key,
+    validationResult: validation,
+  }));
+};
+
+const extractFormRecordErrors = (
+  validationResults: ValidationResult[]
+): ValidationResult[] =>
+  validationResults.filter(fvr => fvr.key === globalFormValidationId);
+
+const removeNotValidValidationResults = (
+  validationResults: ValidationResult[]
+): ValidationResult[] => {
+  if (validationResults.some(fvr => fvr === undefined || fvr === null)) {
+    console.error(
+      'Some form validators are returning null / undefined, fix this'
+    );
+  }
+  return validationResults.filter(fvr => fvr !== undefined && fvr !== null);
+};
+
+const setEmptyKeysToGlobalKeys = (
+  validationResults: ValidationResult[]
+): ValidationResult[] =>
+  validationResults.map(validationResult =>
+    validationResult.key
+      ? validationResult
+      : {
+          ...validationResult,
+          key: globalFormValidationId,
+        }
+  );
+
+const removeValidationsThatSucceeded = (
+  validationResults: ValidationResult[]
+) => validationResults.filter(validationResult => !validationResult.succeeded);
+
+const cleanupValidationResultCollection = (
+  validationResults: ValidationResult[]
+): ValidationResult[] => {
+  let collectionProcessed = removeNotValidValidationResults(validationResults);
+  collectionProcessed = removeValidationsThatSucceeded(collectionProcessed);
+  // TODO check why this is needed
+  // Does it mean global validation arrive here with an empty key ''?
+  // then we map it to global?
+  collectionProcessed = setEmptyKeysToGlobalKeys(collectionProcessed);
+
+  return collectionProcessed;
+};
+
+export const buildFormValidationResult = (
+  validationResults: ValidationResult[]
+): FormValidationSummary => {
+  // TODO: [Dicussion needed here] Should we remove as well validation that had succeeded?
```

Comment:
It depends we want to use it. If we are going to use spread operator to update only the failed values like:

```
interface LoginErrors {
  login: FieldValidationResult;
  password: FieldValidationResult;
}

...

const [loginErrors, setLoginErrors] = React.useState();

validateForm(vm).then(formResult => {
   setLoginErrors({
      ...loginErrors,
      ...formResult.fieldErrors,
   })
})

```

Reply:
We won't be able to use spread operator, since a validation that failed can turn into a successful one,  but I think it can be less prone to errors, not sure as well when showing messages.

Likely we will have to open another branch to fine tune this once we have all the pieces togheter